docs/language-usage.rst

```diff
@@ -0,0 +1,354 @@
+:attr:`Document.HTML <gcloud.language.document.Document.HTML>`.
+
+In addition to supplying the text / HTML content, a document can refer
+to content stored in `Google Cloud Storage`_:
+
+  .. code-block:: python
+
+     >>> document = client.document(bucket='my-text-bucket',
+     ...                            blob='sentiment-me.txt')
+     >>> document.gcs_url
+     'gs://my-text-bucket/sentiment-me.txt'
+
+.. _analyzeEntities: https://cloud.google.com/natural-language/reference/rest/v1beta1/documents/analyzeEntities
+.. _analyzeSentiment: https://cloud.google.com/natural-language/reference/rest/v1beta1/documents/analyzeSentiment
+.. _annotateText: https://cloud.google.com/natural-language/reference/rest/v1beta1/documents/annotateText
+.. _Document: https://cloud.google.com/natural-language/reference/rest/v1beta1/Document
+.. _Google Cloud Storage: https://cloud.google.com/storage/
+
+Analyze Entities
+----------------
+
+The :meth:`~gcloud.language.document.Document.analyze_entities` method
+finds named entities (i.e. proper names) in the text, entity types,
+salience (prominence), mentions for each entity, and other properties.
+
+  .. code-block:: python
+
+     >>> text_val = ("Michelangelo Caravaggio, Italian painter, is "
+     ...             "known for 'The Calling of Saint Matthew'.")
+     >>> document = client.document(text_val)
+     >>> document.analyze_entities()
+     {
+       'entities': [
+         {
+           'name': 'Michelangelo Caravaggio',
+           'type': 'PERSON',
+           'metadata': {
+             'wikipedia_url': 'http://en.wikipedia.org/wiki/Caravaggio'
+           },
+           'salience': 0.75942981,
+           'mentions': [
+             {
+               'text': {
+                 'content': 'Michelangelo Caravaggio',
+                 'beginOffset': 0
+               }
+             }
+           ]
+         },
+         {
+           'name': 'Italian',
+           'type': 'LOCATION',
+           'metadata': {
+             'wikipedia_url': 'http://en.wikipedia.org/wiki/Italy'
+           },
+           'salience': 0.20193423,
+           'mentions': [
+             {
+               'text': {
+                 'content': 'Italian',
+                 'beginOffset': 25
+               }
+             }
+           ]
+         },
+         {
+           'name': 'The Calling of Saint Matthew',
+           'type': 'WORK_OF_ART',
+           'metadata': {
+             'wikipedia_url': 'http://en.wikipedia.org/wiki/index.html?curid=2838808'
+           },
+           'salience': 0.03863598,
+           'mentions': [
+             {
+               'text': {
+                 'content': 'The Calling of Saint Matthew',
+                 'beginOffset': 56
+               }
+             }
+           ]
+         }
+       ],
+       'language': 'en'
+     }
+
+In addition, the encoding of the text in the document can be
+specified in the request. By default, the UTF-8 is used:
```

Comment:
So what's happening here is somewhat subtle. The string that's sent across the wire, I believe, is encoded as UTF8 regardless of the encoding the language uses. However, the API needs to calculate offsets into the string for the entities/etc it detects, which will differ depending on the codec. For example, given the string `1😶2`, in UTF8, the index for `2` is 5. In UTF16, it's 3. In UTF32, it's 2. This is because the index into the string is an index into an array of entries of the smallest 'character' size - 8-bits for UTF8, 16-bits for UTF16, 32-bits for UTF32. So double-byte characters will take up two slots in a UTF8-encoded string, and one slot in a UTF16-encoded string; and triple-byte characters will take up three slots in a UTF8-encoded string, and two slots in a UTF16-encoded string; etc.

So the API needs to know what encoding to calculate the offsets for.

python2 `string` objects are basically byte arrays that are encoded in ascii by default. This barfs on unicode:

```python
>>> a = u'1😶2'
>>> str(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\U0001f636' in position 1: ordinal not in range(128)
```

You can explicitly specify the encoding for unicode characters, for example to UTF8:

```python
>>> a.encode('utf8')
'1\xf0\x9f\x98\xb62'
>>> a.encode('utf8')[5]
'2'
```
in which case you would pass UTF8 to the API.

But if you're using `unicode` objects (ie not explicitly using a particular encoding), you get either UTF32 or UTF16:

```python
>>> sys.maxunicode
1114111
>>> a[2]
u'2'
```
depending on what your python interpreter is compiled as.

Does that make sense? I go through a bunch of examples [here](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/language/api/analyze_test.py#L63) if that would help.